### PR TITLE
Refactor nameserver lookup to an SPI interface

### LINKED ISSUE #6: Android O support
Since Android O removed net.dns1/2/3/4, now the dns look up dose not work on Android O devices, could you introduce Android O support?

### PR DESCRIPTION
**Pull request by ibauersachs**
Closes #6, #8, #55, #57, #68, #80

### CODE REVIEW

File: src/main/java/org/xbill/DNS/config/AndroidResolverConfigProvider.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.xbill.DNS.ResolverConfig;
+import org.xbill.DNS.SimpleResolver;
+
+/**
+ * Resolver config provider for Android. Contrary to all other providers, this provider needs a
+ * context to operate on which must be set by calling {@link #setContext(Object)}.
+ *
+ * <p>If you are developing for Android, consider implementing your own {@link
+ * ResolverConfigProvider} that listens to network callbacks and properly refreshes on link changes.
```
Comment (camhart) [inline src/main/java/org/xbill/DNS/config/AndroidResolverConfigProvider.java]: Can you define "link changes" further?  Does that just mean network changes?

Comment (ibauersachs) [reply src/main/java/org/xbill/DNS/config/AndroidResolverConfigProvider.java]: Yes, I'm not developing for Android, so I can only assume things from reading the docs. `ConnectivityManager.registerNetworkCallback` and `registerDefaultNetworkCallback` seem to be interesting APIs.

**Comment by aaronjwood** [comment]
Will this be released soon? I'm another victim of Android 8+ changes but it looks like this addresses #6

**Comment by ibauersachs** [comment]
I don't have an ETA yet. It would help if you could test the current master branch and provide feedback. But as mentioned in the Javadoc of the Android provider, it would be best if you properly handle Androids' network callbacks in your own application.

**Comment by aaronjwood** [comment]
Ok thanks.

**Pull request merged**